### PR TITLE
VMware: Improve vmware_vmkernel

### LINKED ISSUE #43351: Direct VMKernel Adapter Creation on a dvSwitch portgroup
##### SUMMARY
I'm unable to create VMKernel adapters directly on a dvSwitch portgroup using vmware_vmkernel.

To get a VMKernel adapter created on a dvSwitch I have to do the following;

1. Create a temporary vSwitch using **vmware_vswitch** on the target ESXi host
2. Create a temporary portgroup created using **vmware_portgroup** on the target ESXi host
3. A VMKernel adapter on the temporary portgroup using **vmware_vmkernel** on the taget ESXi host
4. Migrate the VMKernel to the dvSwitch using **vmware_migrate_vmk** 

Deployment and checks will be simplified if vmware_vmkernel can accomodate both switch types as I will be deploying to large clusters of ESXi servers

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

vmware_vmkernel.py

##### ANSIBLE VERSION
```
ansible 2.6.1
  config file = /root/code/vmware-ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### CONFIGURATION

DEFAULT_MANAGED_STR(/root/code/vmware-ansible/ansible.cfg) = This file is managed by Ansible
HOST_KEY_CHECKING(/root/code/vmware-ansible/ansible.cfg) = False
RETRY_FILES_ENABLED(/root/code/vmware-ansible/ansible.cfg) = False

##### OS / ENVIRONMENT
Target is vSphere 6.7, ESXi and vCenter

##### STEPS TO REPRODUCE
The ESXi host is already attached to the dvSwitch "dvs_vmotion" which has the portgroup "vmotion"

```
- name: 'Create a test vmk1'
  vmware_vmkernel:
    hostname: '{{ client_vcenter }}'
    username: '{{ client_vcenter_username }}'
    password: '{{ client_vcenter_password }}'
    esxi_hostname: '{{ ansible_host }}'
    vswitch_name: 'dvs_vmotion'
    portgroup_name: 'vmotion'
    enable_vmotion: True
    vlan_id: 0
    mtu: 9000
    network:
      type: 'static'
      ip_address: '192.168.1.2'
      subnet_mask: '255.255.255.0'
    state: 'present'
    validate_certs: '{{ validate_certs }}'
```

##### EXPECTED RESULTS
The VMKernel adapter would be created on the portgroup on the dvSwitch

##### ACTUAL RESULTS

```
<<ESXI HOSTNAME>> EXEC /bin/sh -c 'echo ~root && sleep 0'
<<ESXI HOSTNAME>> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1532688960.84-255685453744681 `" && echo ansible-tmp-1532688960.84-255685453744681="` echo /root/.ansible/tmp/ansible-tmp-1532688960.84-255685453744681 `" ) && sleep 0'
Using module file /usr/lib/python2.7/site-packages/ansible/modules/cloud/vmware/vmware_vmkernel.py
<<ESXI HOSTNAME>> PUT /root/.ansible/tmp/ansible-local-2808hFnlDs/tmp3damAu TO /root/.ansible/tmp/ansible-tmp-1532688960.84-255685453744681/vmware_vmkernel.py
<<ESXI HOSTNAME>> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1532688960.84-255685453744681/ /root/.ansible/tmp/ansible-tmp-1532688960.84-255685453744681/vmware_vmkernel.py && sleep 0'
<<ESXI HOSTNAME>> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1532688960.84-255685453744681/vmware_vmkernel.py && sleep 0'
<<ESXI HOSTNAME>> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1532688960.84-255685453744681/ > /dev/null 2>&1 && sleep 0'
fatal: [<HOSTNAME>]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "enable_ft": null,
            "enable_mgmt": null,
            "enable_vmotion": true,
            "enable_vsan": null,
            "esxi_hostname": "<ESXI HOSTNAME>",
            "hostname": "<VCENTER HOSTNAME>",
            "ip_address": null,
            "mtu": 9000,
            "network": {
                "ip_address": "192.168.1.2",
                "subnet_mask": "255.255.255.0",
                "type": "static"
            },
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": 443,
            "portgroup_name": "vmotion",
            "state": "present",
            "subnet_mask": null,
            "username": "<ACCOUNT NAME>",
            "validate_certs": false,
            "vlan_id": 0,
            "vswitch_name": "dvs_vmotion"
        }
    },

```

### LINKED ISSUE #48706: Add netStackInstanceKey to vmware_vmkernel
##### SUMMARY
For new interfaces to be created it would be nice to have the option to specify the netStack to be used. 
##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
vmware_vmkernel

##### ADDITIONAL INFORMATION
netStackInstanceKey defaults to TCP/IP but in some cases it makes sense to have a vMotion netStack. However, this can't be changed afterwards without deletion and recreation of the vNic.

```yaml
- name: Create VMKernel for vMotion with vMotion netStack
      vmware_vmkernel:
        hostname: '{{ vcenter_host }}'
        username: '{{ vcenter_user }}'
        password: '{{ vcenter_pass }}'
        esxi_hostname: '{{ item }}'
        validate_certs: yes
        enable_vmotion: yes
        mtu: 1500
        netstack: vmotion
        dvswitch_name: vDS-Managment
        portgroup_name: PG-vMotion
        network:
          type: 'static'
          ip_address: "{{ esxi_hosts[item]['vmotion_ip'] }}"
          subnet_mask: "{{ esxi_hosts[item]['vmotion_mask'] }}"
        state: present
      loop: '{{ esxi_hosts.keys() }}'
      loop_control:
        label: '{{ item }}'
```

### PR DESCRIPTION
**Pull request by ckotte**
##### SUMMARY
* Update description and examples
  * It's not necessary to describe that the VMkernel devices are called 'vnic' internally
  * Especially because a VM nic is called 'vnic' as well
* Add check mode support
* Show settings that were updated / will be updated in the verbose message
* Add default gateway configuration (Check box: 'Override default gateway for this adapter')
* Remove unused option `vlan_id`
* Add vDS support
  * Create port on vDS (Replaces PR https://github.com/ansible/ansible/pull/45636; Fixes https://github.com/ansible/ansible/issues/43351 & #50989)
  * Migrate port from vSS to vDS, or vice versa (Replaces PR https://github.com/ansible/ansible/pull/45636)
  * I didn't get a response from the (other) PR owner and I needed to implement vmkernel ports on vDS in my environment. I also think that we should use `vswitch_name` and `dvswitch_name` and `device_name` should be optional instead of using `is_dvs_switch` and make `device_name` a required option.
* Add TCP/IP stack configuration (fixes #48706)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
vmware_vmkernel

##### ANSIBLE VERSION
```
ansible 2.6.3
  config file = /root/ansible-vmware/ansible.cfg
  configured module search path = [u'/root/ansible-vmware/library']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### ADDITIONAL INFORMATION
Before:
```
TASK [esxi : test] ***************************************************************************************************************************************************************************************************************************
changed: [esxi_host -> localhost] => {"changed": true, "result": "vmk0"}
```
After:
```
TASK [esxi : test] ***************************************************************************************************************************************************************************************************************************
changed: [esxi_host -> localhost] => {"changed": true, "changed_msg": "VMkernel Adapter 'vmk0' updated: MTU (9000) and vMotion (False)", "device": "vmk0"}
```

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_vmkernel.py [hunk 1]
```diff
@@ -374,90 +333,188 @@ def host_vmk_update(self):
         Returns: NA
 
         """
-        results = dict(changed=False, result='')
-        vnic_config = vim.host.VirtualNic.Specification()
-        ip_spec = vim.host.IpConfig()
-        if self.network_type == 'dhcp':
-            ip_spec.dhcp = True
+        changed = False
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vmkernel.py]:
```suggestion
        changed = changed_settings = changed_services = changed_service_vmotion = changed_service_mgmt = changed_service_ft = changed_service_vsan = changed_service_prov = changed_service_rep = changed_service_rep_nfc = False
```

File: lib/ansible/modules/cloud/vmware/vmware_vmkernel.py [hunk 2]
```diff
@@ -374,90 +333,188 @@ def host_vmk_update(self):
...
+                                              "does not exist : %s" % to_native(invalid_arg.msg))
+                # TODO: delete
+                # except Exception as e:
+                #     self.module.fail_json(msg="Failed to add vmk due to general "
+                #                               "exception : %s" % to_native(e))
+
+            if changed_services:
+                vnic_manager = self.esxi_host_obj.configManager.virtualNicManager
+
+                if changed_service_vmotion and self.enable_vmotion:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vmkernel.py]:
```suggestion
                if changed_service_vmotion:
                   operation = 'select' if self.enable_vmotion else 'deselect'
                   self.set_service_type(vnic_manager=vnic_manager, vmk=self.vnic, service_type='vmotion',
                                          operation=operation)
                        
```

Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_vmkernel.py]: same for below.

File: lib/ansible/modules/cloud/vmware/vmware_vmkernel.py (hunk 1 above)
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_vmkernel.py]: done

File: lib/ansible/modules/cloud/vmware/vmware_vmkernel.py (hunk 2 above)
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_vmkernel.py]: Looks much cleaner. Thx.

**Comment by ckotte** [comment]
@Akasurde @dagwieers 
> ```
> lib/ansible/modules/cloud/vmware/vmware_vmkernel.py:0:0: E324 Value for "default" from the argument_spec ({'type': 'static', 'device': None, 'ip_address': None, 'subnet_mask': None, 'default_gateway': None}) for "network" does not match the documentation ({'type': 'static', 'device': 'None', 'ip_address': 'None', 'subnet_mask': 'None', 'default_gateway': 'None'})
> ```

Any idea how to fix this documentation error?

**Comment by Akasurde** [comment]
@mattclay I added `vmware_vmkernel` in ignore list with 324 as suboption validation is not supported right now. Let me know if it is ok, so that I can merge this PR. Thanks.

**Comment by ckotte** [comment]
@Akasurde 
> I added `vmware_vmkernel` in ignore list with 324 as suboption validation is not supported right now. Let me know if it is ok, so that I can merge this PR. Thanks.

I fixed the documentation error. Looks like you cannot use default=None since the documentation doesn't accept it.

**Comment by Akasurde** [comment]
@pdellaert @dericcrago @Im0 @tchernomax @jeking3 @Tomrrow9 Could you please review this ?

**Review by Akasurde — APPROVED** [review]
LGTM

**Comment by dandunckelman** [comment]
@ckotte I just tried using this, but I'm getting this error:

    Portgroup 'pg1' not found on vDS 'dvs1'

If I add that VMK thru the vCenter UI, then it reports what I would expect:

    VMkernel Adapter already configured properly

Your tests don't cover using `dvswitch_name`, which I assume is a limitation of the vcsim.

I believe this issue is already described in #50989.

Please let me know if there's more info I can supply to help.

**Comment by ckotte** [comment]
> Please let me know if there's more info I can supply to help.

@dandunckelman I need more information to troubleshoot this.

I just execute a function with the switch object and the portgroup name. The switch was obviously found, but the function cannot find a valid portgroup.

```
self.dv_switch_obj = find_dvs_by_name(self.content, self.vds_name)
if not self.dv_switch_obj:
    module.fail_json(msg="vDS '%s' not found" % self.vds_name)
self.port_group_obj = find_dvspg_by_name(self.dv_switch_obj, self.port_group_name)
if not self.port_group_obj:
    module.fail_json(msg="Portgroup '%s' not found on vDS '%s'" % (self.port_group_name, self.vds_name))
```

```
def find_dvspg_by_name(dv_switch, portgroup_name):

    portgroups = dv_switch.portgroup

    for pg in portgroups:
        if pg.name == portgroup_name:
            return pg

    return None
```

**Comment by gundalow** [comment]
Marking as WIP till the above is worked out (which will remove `label/shipit`)

**Comment by dandunckelman** [comment]
@ckotte correct. As described in #50989, this appears to be the way ESX operates when it comes to VDS.

I ran the same code you use against my vcenter, and the portgroup that I wanted to add was in the list. But when run directly against the esx host, it only showed portgroups that were already connected. As a result, the `find_dvspg_by_name` method will always `return None` when trying to create a VMK with a portgroup that isn't already on the host.

Even running `esxcli network vswitch dvs vmware list` only lists the connected VDS and the connected portgroups.

I wish I could find documentation to confirm my findings, but I haven't had any luck.

**Comment by ckotte** [comment]
> But when run directly against the esx host, it only showed portgroups that were already connected.

What do you mean with "already connected distributed portgroup"? You can only see distributed portgroups with connected VMs or VMkernel adapters?

I think you have to authenticate to a vCenter and not to an ESXi host if you want to create a new VMkernel on a distributed portgroup. Otherwise, we would need to add a new parameter vcenter that's only required if the connection is local and if a DVS is used. However, I would rather add a comment in the module notes that a vCenter connection is necessary if a DVS is involved or add code that aborts the execution if it's a local ESXi connection.

**Comment by dandunckelman** [comment]
"already connected" meaning a VMK already added to the host.

I think you have to add a param similar to what's done in https://docs.ansible.com/ansible/latest/modules/vmware_dvs_host_module.html#vmware-dvs-host-module

Because I need to be able to state which host receives the VMK while using my vcenter to find the necessary portgroup. So, AFAIK, you'd need both:
- `esxi_hostname` for the host receiving the VMK
- `hostname` for the hostname or IP address of the vcenter

**Comment by ckotte** [comment]
Why not just connecting to the vCenter? The DVS must be configured via vCenter anyway.BTW. I would also need to pass the vCenter credentials since I need to authenticate to query the portgroups.Instead of doing this in the module, this should be done in the playbook.However, I need to change the examples because they use esxi_hostname when using a DVS... I will also add a note to the description...
-------- Original message --------From: Dan Dunckel <notifications@github.com> Date: 2/13/19  15:23  (GMT+01:00) To: ansible/ansible <ansible@noreply.github.com> Cc: Christian Kotte <christian.kotte@gmx.de>, Mention <mention@noreply.github.com> Subject: Re: [ansible/ansible] [WIP] VMware: Improve vmware_vmkernel (#47270) "already connected" meaning a VMK already added to the host.
I think you have to add a param similar to what's done in https://docs.ansible.com/ansible/latest/modules/vmware_dvs_host_module.html#vmware-dvs-host-module
Because I need to be able to state which host receives the VMK while using my vcenter to find the necessary portgroup. So, AFAIK, you'd need both:

esxi_hostname for the host receiving the VMK
hostname for the hostname or IP address of the vcenter


—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/ansible/ansible","title":"ansible/ansible","subtitle":"GitHub repository","main_image_url":"https://github.githubassets.com/images/email/message_cards/header.png","avatar_image_url":"https://github.githubassets.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/ansible/ansible"}},"updates":{"snippets":[{"icon":"PERSON","message":"@dandunckelman in #47270: \"already connected\" meaning a VMK already added to the host.\r\n\r\nI think you have to add a param similar to what's done in https://docs.ansible.com/ansible/latest/modules/vmware_dvs_host_module.html#vmware-dvs-host-module\r\n\r\nBecause I need to be able to state which host receives the VMK while using my vcenter to find the necessary portgroup. So, AFAIK, you'd need both:\r\n- `esxi_hostname` for the host receiving the VMK\r\n- `hostname` for the hostname or IP address of the vcenter"}],"action":{"name":"View Pull Request","url":"https://github.com/ansible/ansible/pull/47270#issuecomment-463216787"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/ansible/ansible/pull/47270#issuecomment-463216787",
"url": "https://github.com/ansible/ansible/pull/47270#issuecomment-463216787",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by dandunckelman** [comment]
@ckotte If you have a working playbook that does just that, then please share.

I'm struggling to see how that could work.

**Comment by ckotte** [comment]
Just like this:

```
- name: Configure vMotion VMkernel adapter
  vmware_vmkernel:
    hostname: "{{ vcenter_server if esxi_connect_via_vcenter else inventory_hostname }}"
    username: "{{ vcsa_username if esxi_connect_via_vcenter else esxi_username }}"
    password: "{{ vcsa_password if esxi_connect_via_vcenter else esxi_password }}"
    esxi_hostname: "{{ inventory_hostname }}"
    vswitch: "{{ dvswitch_vmotion }}"
    portgroup: "{{ dvportgroup_vmotion }}"
    mtu: "{{ esxi_mtu }}"
    enable_vmotion: True
    network:
      type: static
      ip_address: "{{ vmotion_ipv4 }}"
      subnet_mask: "{{ esxi_vmotion_network.netmask }}"
    state: present
    validate_certs: "{{ validate_vcenter_certs }}"
  ignore_errors: "{{ esxi_ignore_errors }}"
  delegate_to: localhost
```

You have to create proper group vars structure depending on your environment. For example, I grouped the ESXi hosts into clusters and I specified the vCenter server FQDN in this file. The credentials can be defined in the same file or on another level.

I usually authenticate via vCenter, but I can authenticate locally as well via an extra variable that I can override. However, this wouldn't work with new VMkernel adapters that need to connect to dvPortgroups.

**Comment by dandunckelman** [comment]
@ckotte how would that work when [this PR's vmware_vmkernel module](https://github.com/ansible/ansible/blob/b2c144ee234a3b38a6319ff00e0355a1188bc8bc/lib/ansible/modules/cloud/vmware/vmware_vmkernel.py) never uses the `hostname` param?

> However, this wouldn't work with new VMkernel adapters that need to connect to dvPortgroups.

Is that the problem described in https://github.com/ansible/ansible/issues/43351?
- That issue best describes my problem
- The workaround listed in the issue description is *exactly* what I have to do ATM
- The [last comment](https://github.com/ansible/ansible/issues/43351#issuecomment-430625572) was you saying that this PR fixes it, but [as I described above](https://github.com/ansible/ansible/pull/47270#issuecomment-463197165):
  - When trying to add a VMK using a dvPortgroup to a host, the call to `find_dvspg_by_name` ONLY returns the list of dvPortgroups already added to the host via other VMKs.

I feel like we're talking on different wavelengths. Please let me know how I can clarify anything. I want this to work for us, but ATM, the bottom line is this: **this PR does not solve the problem described in https://github.com/ansible/ansible/issues/43351**.

**Comment by ckotte** [comment]
> @ckotte how would that work when [this PR's vmware_vmkernel module](https://github.com/ansible/ansible/blob/b2c144ee234a3b38a6319ff00e0355a1188bc8bc/lib/ansible/modules/cloud/vmware/vmware_vmkernel.py) never uses the `hostname` param?

`hostname`, `username`, and `password` are used by every module to authenticate. They are inherited..

> * When trying to add a VMK using a dvPortgroup to a host, the call to `find_dvspg_by_name` ONLY returns the list of dvPortgroups already added to the host via other VMKs.

Yes. Already discussed with you many times. It cannot find all dvPortgroups if not connected to vCenter... You already said it works if connected to vCenter.

> **this PR does not solve the problem described in #43351**.

Of course it does. You just don't get the point. #43351 uses vCenter to authenticate as well. I can create VMkernel adapters if I authenticate to vCenter and NOT ESXi... I did that many times already...

> * The workaround listed in the issue description is _exactly_ what I have to do ATM

Then please use this workaround if it helps you and if don't want to authenticate to vCenter. Why do you need to pollute my PR with this??? Open a new issue if you still don't understand

**Comment by dandunckelman** [comment]
@ckotte 

> `hostname`, `username`, and `password` are used by every module to authenticate. They are inherited..

This! 1000x this. I had NO idea.



I was getting hung up on the fact that I wasn't seeing `hostname` in the code for the `vmware_vmkernel` module. A quick scan of the code brought me to [lib/ansible/module_utils/vmware_rest_client.py](https://github.com/ansible/ansible/blob/64a6dcdd1df64624083880b209f09548f4ec97b0/lib/ansible/module_utils/vmware_rest_client.py), which I believe does exactly what you're saying.

Regardless, I reworked my playbook to use `vmware_vmkernel` as you described, and everything worked as expected!

I feel SO stupid. I'm terribly sorry for burning a bunch of time on this, when I had a simple misunderstanding. Thanks for the PR and your patience in helping me see my mistake.

**Comment by ckotte** [comment]
> I feel SO stupid. I'm terribly sorry for burning a bunch of time on this, when I had a simple misunderstanding. Thanks for the PR and your patience in helping me see my mistake.

No problem. I'm happy to help. I just got a bit annoyed at the end ;)

**Comment by Akasurde** [comment]
@ckotte Ping me when you are done with your changes and we can merge this. Thanks, for the help.

**Comment by ckotte** [comment]
@Akasurde I'm done. I just added a note.

**Comment by ilfox** [comment]
When running the vmware_vmkernel module I'm still getting dvswitch is not supported and vlan_id is still required.

**Comment by dandunckelman** [comment]
@ilfox this hasn't gone out w/ a release yet. I believe it's slated for 2.8.x.

If you need it, then you need to install from the `devel` branch.

Currently, my `requirements.txt` has this:

    -e git://github.com/ckotte/ansible.git@vmware_vmkernel#egg=ansible

but you could use this:

    -e git://github.com/ansible/ansible.git@devel#egg=ansible

**Comment by ilfox** [comment]
Tested and worked on 6.7U1. Thank you!

**Pull request merged**